lib/schedule/data.ex

```diff
@@ -120,181 +187,262 @@ defmodule Schedule.Data do
-      runs
-      |> Map.values()
-      |> Enum.group_by(fn run -> run.service_id end)
 
-    active_services
+    dates
+    |> active_services_on_dates(tables)
     |> Map.new(fn {date, service_ids} ->
       start_time_of_day = Util.Time.time_of_day_for_timestamp(start_time, date)
       end_time_of_day = Util.Time.time_of_day_for_timestamp(end_time, date)
 
-      runs_on_date =
-        Enum.flat_map(service_ids, fn service_id ->
-          Map.get(runs_by_service, service_id, [])
-        end)
+      selectors =
+        for service_id <- service_ids do
+          {{:_, :_, service_id, :"$1"}, [], [:"$1"]}
+        end
 
-      active_runs_on_date =
-        Enum.filter(runs_on_date, fn run ->
-          Run.is_active?(run, start_time_of_day, end_time_of_day)
-        end)
+      active_runs =
+        runs_table
+        |> :mnesia.dirty_select(selectors)
+        |> Enum.filter(&Run.is_active?(&1, start_time_of_day, end_time_of_day))
 
-      {date, active_runs_on_date}
+      if active_runs == [] do
+        {nil, []}
+      else
+        {date, active_runs}
+      end
     end)
-    |> Helpers.filter_values(fn runs -> runs != [] end)
+    |> Map.delete(nil)
   end
 
-  @spec shapes(t(), Route.id()) :: [Shape.t()]
-  def shapes(%__MODULE__{shapes: shapes}, route_id), do: Map.get(shapes, route_id, [])
-
-  @spec shape_for_trip(t(), Schedule.Trip.id()) :: Shape.t() | nil
-  def shape_for_trip(%__MODULE__{shapes: shapes, trips: trips}, trip_id) do
-    trip = Map.get(trips, trip_id)
+  @spec shapes(tables(), Route.id()) :: [Shape.t()]
+  def shapes(%{shapes: shapes_table}, route_id) do
+    :mnesia.dirty_select(shapes_table, [
+      {{:_, :_, route_id, :"$1"}, [], [:"$1"]}
+    ])
+  end
 
-    if trip != nil do
-      route_shapes = Map.get(shapes, trip.route_id, [])
-      Enum.find(route_shapes, fn shape -> shape.id == trip.shape_id end)
+  @spec shape_for_trip(tables(), Schedule.Trip.id()) :: Shape.t() | nil
+  def shape_for_trip(%{shapes: shapes_table} = tables, trip_id) do
+    with %{route_id: route_id} <- trip(tables, trip_id),
+         [{_, _, _, shape}] <- :mnesia.dirty_index_read(shapes_table, route_id, :route_id) do
+      shape
     else
-      nil
+      _ ->
+        nil
     end
   end
 
-  @spec first_route_pattern_for_route_and_direction(t(), Route.id(), Direction.id()) ::
+  @spec first_route_pattern_for_route_and_direction(tables(), Route.id(), Direction.id()) ::
           RoutePattern.t() | nil
   def first_route_pattern_for_route_and_direction(
-        %__MODULE__{route_patterns: route_patterns},
+        %{route_patterns: route_patterns_table},
         route_id,
         direction_id
       ) do
-    Enum.find(route_patterns, fn route_pattern ->
-      route_pattern.route_id == route_id && route_pattern.direction_id == direction_id
-    end)
+    case :mnesia.dirty_read(
+           route_patterns_table,
+           {route_id, direction_id}
+         ) do
+      [{_, _, route_pattern}] -> route_pattern
```

Comment:
**question:** shouldn't this also match the case where there are multiple route patterns (but still return only the first one)?

Reply:
Since this table is a `:set` there's at most one record for a given `{route_id, direction_id}` pair.

Reply:
Ahh, right, now I understand what you did with the `Enum.reverse` in `save_schedule_data_to_tables/2`.

Reply:
I originally had it the other way (since the table used `route_pattern.id` as the key) but I don't believe Mnesia guarantees that results are returned in the same order that they were inserted. And either way, none of the code cares about route patterns that aren't the first for their route/direction.

Reply:
Makes sense, and we can always go back and change things if we need the other route patterns in some ordered fashion in the future.